clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java

```diff
@@ -156,6 +221,97 @@ else if (key.defaultValue == NO_DEFAULT_VALUE)
         return values;
     }
 
+
+    /**
+     * Validate the provided connector configuration values with the configuration definition
+     * @param connectorConfigs the provided connector configuration values
+     * @return List of Config, each Config contains the updated configuration information given
+     * the current configuration values.
+     */
+    public List<Config> validate(Map<String, String> connectorConfigs) {
+        List<String> configsWithNoParent = getConfigsWithNoParent();
+        Map<String, Config> configValues = new HashMap<>();
+        for (String name: configsWithNoParent) {
+            validate(name, connectorConfigs, configValues, name, null);
+        }
+        return new LinkedList<>(configValues.values());
+    }
+
+
+    private List<String> getConfigsWithNoParent() {
+        Set<String> configs = configKeys.keySet();
+        Set<String> configsWithParent = new HashSet<>();
+
+        for (ConfigDef.ConfigKey configKey: configKeys.values()) {
+            List<String> dependents = configKey.dependents;
+            for (String name: dependents) {
+                configsWithParent.add(name);
+            }
+        }
+
+        List<String> configsWithNoParent = new LinkedList<>();
+        for (String name: configs) {
+            if (!configsWithParent.contains(name)) {
+                configsWithNoParent.add(name);
+            }
+        }
+        return configsWithNoParent;
+    }
+
+    private void validate(String name, Map<String, String> connectorConfigs, Map<String, Config> configs, String parentName, Object parentValue) {
+        ConfigKey key = configKeys.get(name);
+
+        Config config;
+        if (configs.containsKey(name)) {
+            config = configs.get(name);
+        } else {
+            config = new Config(name);
+        }
+
+        Object value;
+        if (connectorConfigs.containsKey(key.name)) {
+            value = parseType(key.name, connectorConfigs.get(key.name), key.type);
+        } else if (key.defaultValue == NO_DEFAULT_VALUE) {
+            config.addErrorMessage("Missing required configuration \"" + key.name + "\" which has no default value.");
+            value = null;
+        } else {
+            value = key.defaultValue;
+        }
+        if (key.validator != null) {
+            key.validator.ensureValid(key.name, value);
+        }
+
+        config.setValue(value);
+
+        Set<Object> recommendedValues;
+        if (key.recommender != null) {
+            recommendedValues = key.recommender.validValues(name, parentName, parentValue);
+            List<Object> originalRecommendedValues = config.getRecommendedValues();
+
+            if (!originalRecommendedValues.isEmpty()) {
```

Comment:
Not sure I understand what's going on here. Why wouldn't we just use the current recommended values? I guess it has to do with the config being stateful, but if we filter based on the previous recommended values, can't we get ourselves into a situation where no values are recommended, which means none will ever be recommended again since we filter based on the previous set?

It seems like we could just use the current set of values directly. Additionally, using a `Set` may not be ideal here since it'll shuffle the ordering, which could potentially be useful information. Why not just accept a list of recommended values and trust the connector to return correct values?